### ISSUE TITLE
XmlCrefAttributeSyntax with generic types

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#235]
I tried to create a `cref` attribute for an element in an XML documentation element using the following code:

```csharp
TypeSyntax typeSyntax = SyntaxFactory.ParseTypeName("global::System.Lazy<T>");
TypeCrefSyntax cref = SyntaxFactory.TypeCref(typeSyntax);
XmlCrefAttributeSyntax crefAttribute = SyntaxFactory.XmlCrefAttribute(
    SyntaxFactory.XmlName("cref"),
    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken),
    cref,
    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken));
```

When I added this attribute to an element in a code fix operation with the `Simplifier.Annotation` annotation, the resulting code for this specific attribute looked like this:

```
cref="Lazy<T>"
```

I expected the attribute to actually be emitted as:

```
cref="Lazy{T}"
```

:question: How can I produce a correct `cref` attribute for an XML documentation comment from an arbitrary `TypeSyntax`?

**Comment by amcasey** [comment]
Unfortunately, I don't think we have code for doing that anywhere in Roslyn - internal or otherwise.  We've talked about adding a public API for parsing cref syntax, but I don't believe that exists at the moment either (and it doesn't sound like it would solve your problem).

If you're interested in writing one, let me know and we can think about it together.  The syntax for crefs is a bit convoluted.

**Comment by sharwell** [comment]
What I'm looking for is more like `SymbolDisplayFormat.FullyQualifiedDocumentationFormat`.

**Comment by amcasey** [comment]
That seems like a clean way to expose the functionality, but it definitely doesn't exist at the moment.

**Comment by sharwell** [comment]
It looks like the completion provider handles this on its own:

http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Completion/CompletionProviders/CrefCompletionProvider.cs,198

**Comment by amcasey** [comment]
Oh, cool, I guess there is code.  @rchande should be able to tell you more about it.

**Comment by sharwell** [comment]
Oh wow...

1. If you use `SyntaxFactory.ParseTypeName("global::OpenStack.Net.IHttpApiCall{T}")`, it works but the simplifier will not fully simplify the name. If you have `using OpenStack.Net;`, then the result will be `cref="Net.IHttpApiCall{T}"`. This is actually a bug in `ParseTypeName` or in the simplifier which should be addressed. However...
2. I made it work "for real" using the names produced by `SymbolDisplayFormat.FullyQualifiedName`.

```csharp
CrefSyntax cref = ...;
// BEGIN new code
cref = cref.ReplaceTokens(cref.DescendantTokens(), ReplaceBracketTokens);
// END new code
return SyntaxFactory.XmlCrefAttribute(
    SyntaxFactory.XmlName("cref"),
    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken),
    cref,
    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken));
```

The above uses the following helper function. This method rewrites the `Text` property of the angle bracket tokens without changing their `SyntaxKind` or their `ValueText`, which produces the correct generated output code without restricting the input or limiting the simplifier functionality.

```csharp
private static SyntaxToken ReplaceBracketTokens(SyntaxToken originalToken, SyntaxToken rewrittenToken)
{
    if (rewrittenToken.IsKind(SyntaxKind.LessThanToken) && string.Equals("<", rewrittenToken.Text, StringComparison.Ordinal))
        return SyntaxFactory.Token(rewrittenToken.LeadingTrivia, SyntaxKind.LessThanToken, "{", rewrittenToken.ValueText, rewrittenToken.TrailingTrivia);

    if (rewrittenToken.IsKind(SyntaxKind.GreaterThanToken) && string.Equals(">", rewrittenToken.Text, StringComparison.Ordinal))
        return SyntaxFactory.Token(rewrittenToken.LeadingTrivia, SyntaxKind.GreaterThanToken, "}", rewrittenToken.ValueText, rewrittenToken.TrailingTrivia);

    return rewrittenToken;
}
```

**Comment by sharwell** [comment]
See here for the change:

https://github.com/sharwell/OpenStackNetAnalyzers/commit/67bbd18bdef7f8a364c7e7db15d4b33e6731f62b

**Comment by amcasey** [comment]
I'm a little surprised that ParseTypeName handles curly brackets.  Are you sure it's not just truncating?

Also, you might want to see what happens if you replace "T" with "int", since that's not legal in crefs.

**Comment by sharwell** [comment]
My helper class `XmlSyntaxFactory` assumes that the `CrefSyntax` you give it is a reference to something that *can* be represented in an XML documentation comment. While it's not possible to represent `Tuple<int>`, it is possible for `Tuple<int>` to appear as an argument to a method you reference, in which case `Tuple{T1}` would not be a valid substitution. For example:

```csharp
/// <summary>
/// Valid: <see cref="Foo(Tuple{int})"/>
/// Not valid: <see cref="Foo(Tuple{T1})"/>
/// </summary>
public void Foo(Tuple<int> value)
{
}
```

**Comment by sharwell** [comment]
> I'm a little surprised that ParseTypeName handles curly brackets. Are you sure it's not just truncating?

The output when I passed the curly brackets to ParseTypeName was a valid, but not fully simplified, `cref` attribute in the XML.

**Comment by amcasey** [comment]
Sounds like you have a good handle on things.  At this point, your remaining concerns are around the availability of APIs to help with this scenario?

**Comment by sharwell** [comment]
Yes, it seems like this conversion should take place in `SyntaxFactory.TypeCref` (and related).

The overall issue of "crazy complexity" of the `SyntaxFactory` methods for XML documentation comments is a separate request (#218).

**Comment by Pilchie** [comment]
So, is there anything left of this request that isn't covered by #218?

**Comment by sharwell** [comment]
@Pilchie They could be combined, but you can see in #218 that I assumed it's safe to pass the result of `SyntaxFactory.TypeCref` straight into the documentation comment factory methods. This is not currently the case due to the fact that the factory does not translate between the generic forms of the input. Since this is more of a problem with the implementation of `SyntaxFactory.XmlCrefAttribute` than with the public API surface, they would probably best be dealt with separately.

**Comment by Pilchie** [comment]
Thanks Sam - moving over to the compiler as this is a Syntax level API.
